crates/ruff/src/rules/flake8_simplify/rules/ast_if.rs

```diff
@@ -627,75 +649,63 @@ pub(crate) fn use_ternary_operator(checker: &mut Checker, stmt: &Stmt, parent: O
+pub(crate) fn if_with_same_arms(checker: &mut Checker, locator: &Locator, stmt: &Stmt) {
+    let Stmt::If(ast::StmtIf { body, elif_else_clauses, .. }) = stmt else {
         return;
     };
 
-    // It's part of a bigger if-elif block:
-    // https://github.com/MartinThoma/flake8-simplify/issues/115
-    if let Some(Stmt::If(ast::StmtIf {
-        orelse: parent_orelse,
-        ..
-    })) = parent
-    {
-        if parent_orelse.len() == 1 && stmt == &parent_orelse[0] {
-            // TODO(charlie): These two cases have the same AST:
-            //
-            // if True:
-            //     pass
-            // elif a:
-            //     b = 1
-            // else:
-            //     b = 2
-            //
-            // if True:
-            //     pass
-            // else:
-            //     if a:
-            //         b = 1
-            //     else:
-            //         b = 2
-            //
-            // We want to flag the latter, but not the former. Right now, we flag neither.
-            return;
+    // Check adjacent `if` or `elif` branches, ignore `else`
+    let mut branches_iter = iter::once((stmt.start(), body.as_slice()))
+        .chain(
+            elif_else_clauses
+                .iter()
+                .filter(|clause| clause.test.is_some())
+                .map(|clause| (clause.start(), clause.body.as_slice())),
+        )
+        .peekable();
+
+    while let Some((first_start, first_body)) = branches_iter.next() {
+        let Some((second_start, second_body)) = branches_iter.peek() else {
+            continue
+        };
+
+        // The bodies must have the same code ...
+        if first_body.len() != second_body.len() {
+            continue;
+        }
+        if !first_body
+            .iter()
+            .zip(second_body.iter())
+            .all(|(stmt1, stmt2)| compare_stmt(&stmt1.into(), &stmt2.into()))
+        {
+            continue;
         }
-    }
 
-    let if_body_pairs = get_if_body_pairs(test, body, orelse);
-    for i in 0..(if_body_pairs.len() - 1) {
-        let (test, body) = &if_body_pairs[i];
-        let (.., next_body) = &if_body_pairs[i + 1];
-        if compare_body(body, next_body) {
-            checker.diagnostics.push(Diagnostic::new(
-                IfWithSameArms,
-                TextRange::new(
-                    if i == 0 { stmt.start() } else { test.start() },
-                    next_body.last().unwrap().end(),
-                ),
-            ));
+        // ...and the same statements
+        // TODO(konstin): Don't we have any helper to extract the comments in a specific range?
+        let comments_in_range = |range: TextRange| {
+            checker
+                .indexer
+                .comment_ranges()
+                .iter()
+                .filter(|comment_range| range.contains(comment_range.start()))
+                .map(|comment_range| locator.slice(*comment_range))
+                .collect()
```

Comment:
You can add a helper method to `CommentRanges` that performs a binary search using the start position, then checks if the range is included)